### PR TITLE
resourcemanager: support worker pool

### LINKED ISSUE #42511: Support a worker pool
## Enhancement
We need to support a worker pool that can change workers dynamically by the resource manager.

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #42511

Problem Summary:

### What is changed and how it works?
This PR supports a worker pool that can change workers dynamically by the resource manager.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: resourcemanager/pool/workerpool/workpool_test.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	id int
+}
+
+func (w *MyWorker) HandleTask(task any) {
+	globalCnt.Add(task.(int64))
+	cntWg.Done()
+	logutil.BgLogger().Info("Worker handling task")
+}
+
+func (w *MyWorker) Close() {
```
Comment (hawkingrei) [inline resourcemanager/pool/workerpool/workpool_test.go:41]: ```MyWorker``` -> ```Worker```

File: resourcemanager/pool/workerpool/workerpool.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+type Worker interface {
+	HandleTask(task any)
+	Close()
+}
+
+// WorkerPool is a pool of workers.
+type WorkerPool[T any] struct {
+	name           string
+	numWorkers     int32
+	runningWorkers atomicutil.Int32
```
Comment (tangenta) [inline resourcemanager/pool/workerpool/workerpool.go:37]: Running tasks?

Comment (wjhuang2016) [reply resourcemanager/pool/workerpool/workerpool.go:37]: Yes， it means workers that running the task.

**Review by tangenta — APPROVED** [review]

File: resourcemanager/pool/workerpool/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        "//resourcemanager/util",
+        "//util",
+        "@com_github_sasha_s_go_deadlock//:go-deadlock",
+        "@org_uber_go_atomic//:atomic",
+    ],
+)
+
+go_test(
+    name = "workerpool_test",
+    srcs = [
```
Comment (hawkingrei) [inline resourcemanager/pool/workerpool/BUILD.bazel:19]: Please add 

```
timeout = "short",
flaky = True,
race = "on",
```

and run ```make bazel_prepare```.

File: resourcemanager/pool/workerpool/workerpool.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+func (p *WorkerPool[T]) runAWorker() {
+	p.wg.Run(func() {
+		w := p.createWorker()
+		for {
+			select {
+			case task := <-p.taskChan:
+				p.runningTask.Add(1)
+				w.HandleTask(task)
```
Comment (hawkingrei) [inline resourcemanager/pool/workerpool/workerpool.go:80]: Should we add ```recovery```here to avoid panic?

Comment (wjhuang2016) [reply resourcemanager/pool/workerpool/workerpool.go:80]: let the worker to handle it.

Comment (hawkingrei) [reply resourcemanager/pool/workerpool/workerpool.go:80]: If wokrer forgets to handle it, how to ensure the correctness here.

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**